pulser-core/pulser/operators/operator.py

```diff
@@ -0,0 +1,419 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Classes to store Operators for simulation and measurements."""
+
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections import UserList
+from collections.abc import Collection
+import itertools
+from dataclasses import dataclass
+from numbers import Number
+from typing import Mapping
+
+from pulser.channels import State, STATES_RANK
+from pulser.register import QubitId
+
+PROJECTORS = {"sigma_" + a + b for a in STATES_RANK for b in STATES_RANK}
+
+
+def _check_kron_object(
+    kron_obj: tuple[tuple[QuditString | QuditOperatorString, set[QubitId]]],
+    obj_type: type,
+) -> None:
+    """Check if an object with Kronecker products is correctly defined."""
+    if not isinstance(kron_obj, tuple) or any(
+        [isinstance(operation[0], obj_type) for operation in kron_obj]
+    ):
+        raise ValueError(
+            "The kron_"
+            f"{"qudits" if str(obj_type) == "QuditString" else "operations"}"
+            " should be a tuple of tuples of shape (operator as a "
+            f"{obj_type}, qubit id as an int or str)."
+        )
+    # Check the qubit ids targeted in operations
+    target_qids = [operation[1] for operation in kron_obj]
+    if "global" in target_qids:
+        if len(kron_obj) > 1:
+            raise ValueError(
+                "If a 'global' operation is defined, no other operations "
+                "can be defined."
+            )
+    else:
+        target_qids_list = list(itertools.chain(*target_qids))
+        target_qids_set = set(target_qids_list)
+        if len(target_qids_set) < len(target_qids_list):
+            # Either the qubit id has been defined twice in an operation:
+            for qids in target_qids:
+                if len(set(qids)) < len(qids):
+                    raise ValueError("Duplicate atom ids in argument list.")
+            # Or it was defined in two different operations
+            raise ValueError(
+                "Each qubit can be targeted by only one operation."
+            )
+
+
+@classmethod
+class QuditString:
+    """A 1-qudit state."""
+
+    coefficients: list[Number]
+    states: list[State]
+
+    def __post_init__(self):
+        for state in self.states:
+            if state not in STATES_RANK:
+                raise ValueError(f"State {state} does not exist.")
+
+    def as_tuple(self) -> tuple[tuple[Number, State]]:
+        """Returns tuple of (coefficient, state label)."""
+        return zip(self.coefficients, self.states)
+
+
+@classmethod
+class MultiQuditString:
```

Comment:
This implementation suffers from the same issue we discussed with `Operator`, namely, any state can be written as a `sum` of `kron`, where the `kron` is over basis elements, but not every state can be written as a `kron` of `sum` where the sum is over basis elements. For this reason, I would do something like 

```
QuditString = str

class MultiQuditString:
    """weighted sum of QuditString."""
    coefficients: list[Number]
    states: list[QuditString]
    def as_tuple(self) -> tuple[tuple[Number, QuditString]]:
        pass #too lazy to write the implementation
```
Of course

Reply:
Are QuditString and MultiQuditString really necessary ? I implemented this to implement the __mul__ operation in the current Operator, but in the new implementation I am not sure it is...

Otherwise I completely agree with your comment.

Reply:
So, if I understand your motivation correctly, you were defining these so that the user could have a way to define operators without a backend specific support, just from strings. 
I understand the need for not being dependent on `qutip` to define your operators (let's say you only want to interact with Emu-T, it doesn't make sense to have a `qutip` dependency), but perhaps it would be simpler to define operators with a `numpy` support instead for that case, rather than working out a string based algebra.

Reply:
If I were to implement states - I want to be 100% sure this is needed beforehand. I could do (to keep the spirit of Operator): QuditString (1-qudit state)- TargetedQuditString (multi-qudit state as kron of qudit states) - MultiQuditString (sum of kron) - MultiQudit(ABC) (abstract class like Operator) - NumpyMultiQudit(MultiQudit) (stores a numpy array, makes operation with this array, convert this array into string and from string into this array)

Reply:
I feel like this is getting out a hand, I would suggest we meet and come up with a plan that keeps things as simple as possible while delivering the functionality we are looking for.

Reply:
I think QuditString is too far fetched and should not be implemented, nor the operations on OperatorString. I definitely think we should go with something minimal. I will resume my work on this PR tomorrow.

Reply:
I would like to have an abstract representation for strings, so that users can measure coefficients of the wavefunction via the inner product on states. I don't care about implementations such as `OperatorString*QuditString` very much.